### PR TITLE
Add visual indication of lost logs to UART and RTT backends 

### PR DESCRIPTION
**Pull request by nordic-krch**
Added implementation of new interface added in #11769.
It touches #11763 as it fixes it for UART, RTT backends. native_posix and shell is still not implementing it.

When messages are dropped `~` is printed. Number of printed `~` indicates number of dropped messages, giving rough idea how much buffer should be increased.

### CODE REVIEW

File: subsys/logging/log_backend_uart.c [hunk 1]
```diff
@@ -62,10 +62,24 @@ static void panic(struct log_backend const *const backend)
...
+{
+	int i;
+	struct device *dev = log_output.control_block->ctx;
+	static const char newline_str[] = "\r\n";
+
+	for (i = 0; i < cnt; i++) {
+		uart_poll_out(dev, '~');
+	}
+
+	char_out((u8_t *)newline_str, sizeof(newline_str), dev);
```
Comment (jhedberg) [inline subsys/logging/log_backend_uart.c]: sizeof(newline_str) evaluates to 3, is it intentional that you're also writing the terminating null byte?

Comment (nordic-krch) [reply subsys/logging/log_backend_uart.c]: right, sending null wasn't intentional. Fixed to send sizeof - 1.

**Comment by pfalcon** [comment]
Well, that's definitely a step in the right direction, but why not print clearly: `~Dropped: NN\n` ?

**Comment by pfalcon** [comment]
In particular, why print a thousand of `~`, if 1000 messages were dropped?

**Review by pizi-nordic — COMMENTED** [review]
I agree with @pfalcon, that printing "--- %u messages lost ---" will be better. However I understand, that we should avoid printing too much, as we should use the bandwidth for log messages.

**Comment by nordic-krch** [comment]
ok, i agree that message might be better but it i would add red coloring to it as it must be spotted by the user.

**Comment by nordic-krch** [comment]
please re-review as I've reworked the thing:
- added generic handling of dropped messages in log_output
- backends are using log_output function so output will be uniform across backends
- adding dropped message support to the backend is trivial
- message `--- nn messages dropped ---` is red if coloring is enabled

File: subsys/logging/log_output.c [hunk 2]
```diff
@@ -522,6 +528,24 @@ void log_output_msg_process(const struct log_output *log_output,
...
+{
+	char buf[3];
+	int len;
+	static const char prefix[] = DROPPED_COLOR_PREFIX "--- ";
+	static const char postfix[] =
+			" messages dropped ---\r\n" DROPPED_COLOR_POSTFIX;
+	log_output_func_t outf = log_output->func;
+	struct device *dev = (struct device *)log_output->control_block->ctx;
+
+	cnt = min(cnt, 99);
```
Comment (pfalcon) [inline subsys/logging/log_output.c]: But why? What's wrong with printing 5-digit number?

File: subsys/logging/log_output.c [hunk 3]
```diff
@@ -114,18 +115,23 @@ static int print_formatted(const struct log_output *log_output,
 	return length;
 }
 
-void log_output_flush(const struct log_output *log_output)
+static void buffer_flush(log_output_func_t outf, u8_t *buf, size_t len,
```
Comment (pfalcon) [inline subsys/logging/log_output.c]: I find it strange that function which performs "write" operation is called "flush". "Flush" operation generally doesn't take params (except for maybe timeout), and flushes whatever already accumulated.

Comment (nordic-krch) [reply subsys/logging/log_output.c]: yes, i did rename only function prefix after extracting it from `log_output_flush`. changed to `buffer_write`.

File: subsys/logging/log_output.c (hunk 2 above)
Comment (nordic-krch) [reply subsys/logging/log_output.c]: increased to 4 digits. It's unlikely to see more than 9999 lost messages and even then exact number is not important.

**Review by pfalcon — APPROVED** [review]
Thanks for the updates!

**Comment by pfalcon** [comment]
@jhedberg, @pizi-nordic, @jarz-nordic, @carlescufi: Can you please (re)review this, so we paced forward with logging fixes?

**Review by carlescufi — APPROVED** [review]

**Review by pizi-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**